File: server/db/loadData.js
import { GoogleGenerativeAI } from "@google/generative-ai";
import { DataAPIClient } from "@datastax/astra-db-ts";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import { data } from "../utils/data.js";
import dotenv from 'dotenv';

dotenv.config();

const apiKey = "AIzaSyBsA3-aS6Pbz1-g6x0ZZaUxg3zPyTAtz5A";

const geminiAI = new GoogleGenerativeAI(apiKey);

const model = geminiAI.getGenerativeModel({ model: "models/text-embedding-004" });

const client = new DataAPIClient('AstraCS:KrKELLMinpUAKAaLKyYTusiD:f8769903e7e7fde223efa698b31551c8f3eb5f6b8cfa9ef7b41b2e041d97399d');
const db = client.db('https://25d21ac4-184f-4816-a1f9-8b20c55ee198-us-east-2.apps.astra.datastax.com');

const splitter = new RecursiveCharacterTextSplitter({
  chunkSize: 1000,
  chunkOverlap: 200,
});

async function createCollection() {
  try {
    await db.createCollection("portfolio", {
      vector: {
        dimension: 1536,
      },
    });
  } catch (error) {
    console.log("Collection Already Exists", error);
  }
}

async function loadData() {
  const collection = db.collection("portfolio");
  for await (const { id, info, description } of data) {
    const chunks = await splitter.splitText(description);

    for await (const chunk of chunks) {
      const data = await model.embedContent({
        content: chunk,
      });

      const res = await collection.insertOne({
        document_id: id,
        $vector: data?.embedding,
        info,
        description: chunk,
      });
    }
  }
  console.log("data added");
}

createCollection().then(() => loadData());

File: server/db/loadData.py
from langchain_community.document_loaders import TextLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
import json
from sentence_transformers import SentenceTransformer
import numpy as np
from astrapy import DataAPIClient
import os
from dotenv import load_dotenv

load_dotenv()

# Environment variables
db_token = os.getenv('DB_TOKEN')
db_endpoint = os.getenv('DB_ENDPOINT')

# Initialize the DB client
client = DataAPIClient(db_token)
db = client.get_database_by_api_endpoint(
  db_endpoint
)
print(f"Connected to Astra DB: {db.info().name}")
print(f"Connected to Astra DB: {db.list_collection_names()}")

# Create a collection.
collection = db.create_collection(
    "portfolio",
    dimension=768,
    metric= "cosine",  # or simply "cosine"
    check_exists=False,
)
print(f"* Collection: {collection.full_name}\n")

# Read and load the JSON data from the file
with open('./utils/data.json', 'r') as file:
    text_list = json.load(file)

# Spiltter
r_splitter = RecursiveCharacterTextSplitter(
    separators = ["\n\n", "\n", " "],
    chunk_size = 200,
    chunk_overlap  = 0,
    length_function = len
)

# Encoder
encoder = SentenceTransformer("all-mpnet-base-v2")

for item in text_list:
        id = item['id']
        title = item['title']
        description = item['description']
        
        chunks = r_splitter.split_text(description)
        
        for chunk in chunks:
            vector = encoder.encode(chunk)
            dim=vector.shape
            print(dim)

            try:
                insertion_result = collection.insert_one({
                    "_id": id,
                    "title": title,
                    "$vector": vector,
                    "description": chunk
                })
                print(f"* Inserted {len(insertion_result.inserted_ids)} items.\n")
            except:
                print("* Documents found on DB already. Let's move on.\n")
